### ISSUE TITLE
[SSO] Allow a username that is different from UID

### ISSUE THREAD

**Issue by defnull** [mastodon/mastodon#21296]
### Pitch

Allow SSO users to choose their own `username` on first login and make it independent from the `uid` used to map SSO identities.

### Motivation

I'm currently trying to figure out how to setup a Mastodon instance for a large academic organization with hundreds of thousands of potential users.

Our SSO has a unique and stable ID for each account, but this ID is numeric and unsuitable for public profile names. No one wants to remember long numbers to be able to tag people.

We also have a `preferred_username` but this is generated from the actual name plus some ugly suffix to avoid conflicts (e.g. `j.doe12`). People may not want to use that as their public social identity. Even more challenging is the fact that these names are unique, but not stable. People marry of have other more personal reasons to change their name, so they usually also want to change their username and email address. The old profile is now inaccessible because Mastodon creates a new `Identity`, `User` and `Account` on login. Since old `preferred_username`s can be re-assigned to new accounts after a while, this also introduces a security risk. All in all, this would be a support nightmare down the road.

By allowing users to choose their own username on first login (similar to how mastodon asks for an e-mail if the SSO did not provide one), we would avoid most of these issues. We could choose a unique and stable SSO attribute to safely identify accounts, and allow users to choose their preferred username instead of being forced to stick with their ugly auto-generated one.

**Comment by ineffyble** [comment]
On a much smaller scale we face this issue with toot.thoughtworks.com (using SAML).

**Comment by echox** [comment]
Partly also mentioned here with another example in the comments: https://github.com/mastodon/mastodon/issues/20672

**Comment by CyberPingU** [comment]
Hello,
are there any news on this issue?
In my case (I work for a local government office), our OIDC providers just give us the preferred name that is our TAX ID: it's really impossibile for us to show the TAX-ID  as username, because it would be an enormous privacy breach (I live in the EU)

**Comment by ThisIsMissEm** [comment]
I'm currently exploring whether we can provide the existing non-SSO registration experience for SSO.

For now, for OIDC providers, you can use `OIDC_UID_FIELD=preferred_username` (or any other value of the User Info Endpoint to set what username value to use from the OIDC provider), if the `uid` value from your SSO provider is a random identifier.

That said, I really wouldn't recommend linking your Mastodon server to government-backed SSO based around Tax IDs.

This is similar as an issue to #22837 and part of #20672

**Comment by defnull** [comment]
> For now, for OIDC providers, you can use `OIDC_UID_FIELD=preferred_username` (or any other value of the User Info Endpoint to set what username value to use from the OIDC provider), if the `uid` value from your SSO provider is a random identifier.

That may be a really bad idea because `preferred_username` is, by definition, not unique and not fixed. If you can choose your own preferred username, you could just use the same as an existing user (or the admin) and log in as that user.

**Comment by defnull** [comment]
This is how I imagine a (backwards compatible) solution:
* `OIDC_UID_FIELD` is used to map SSO identities to Mastodon users, as it is now. It should be a stable and unique attribute that never changes and will never be re-assigned to a different SSO identity (e.g. `sub`).
* (new) `OIDC_USERNAME_FIELD` is used to generate the default username for new users. It defaults to `OIDC_UID_FIELD` but can be changed to a nicer attribute (e.g. `preferred_username`) if the UID field is not suitable for public usernames.
* (new) If `OIDC_USERNAME_FORCED` is `False` (default: `True`) then new users can change their username on first login before the account is created. A form asking new users for username and display name can always be shown, but the username field would be read-only if `OIDC_USERNAME_FORCED` is `True`.

**Comment by ThisIsMissEm** [comment]
> > For now, for OIDC providers, you can use `OIDC_UID_FIELD=preferred_username` (or any other value of the User Info Endpoint to set what username value to use from the OIDC provider), if the `uid` value from your SSO provider is a random identifier.
> 
> That may be a really bad idea because `preferred_username` is, by definition, not unique and not fixed. In many environments you can choose your own preferred username for your SSO account. By choosing the same name as an existing Mastodon user or admin, you could log into Mastodon as that user.
> 
> For a secure SSO flow the Identity should be linked using a unique and stable SSO attribute, and those are usually generated internal IDs not suitable for human readable handles. The user should either be allowed to choose a custom handle on first login, unless the admin configures a _separate_ SSO attribute to use as the handle.

No, that's not how this works. The preferredUsername is just used as the starting value, a unique username is still generated by Mastodon if necessary. No account takeover attacks should be possible through this mechanism.

**Comment by defnull** [comment]
Please correct me if I'm wrong but:

* The value of the SSO attribute defined by `OIDC_UID_FIELD` ends up in omniauth `auth.uid`. If I define `OIDC_UID_FIELD=preferred_username` and my preferred username in SSO is `foo` then `auth.uid` will be `foo`. 
* Mastodon will find-or-create an `Identity` with that `uid`. See https://github.com/mastodon/mastodon/blob/main/app/models/concerns/user/omniauthable.rb#L22 and https://github.com/mastodon/mastodon/blob/main/app/models/identity.rb#L21.
* If the identity exists and `Identity.user` has a value, then that user will be logged in. If I manage to have the same `preferred_username` and thus `auth.uid` and `identity.uid` as an existing user, I will be logged in as that user. This is why `OIDC_UID_FIELD` must point to a stable and unique attribute, or collisions may be possible which leads to an information leak (PMs) or account takeover. Recommending an often user-defineable field like `preferred_username` for thus putpose is a disaster waiting to happen.
* Only if there is no user attached to that Identity yet, Mastodon will assume it's a first time SSO login. It will first search for an existing user with the same email and attach that. See https://github.com/mastodon/mastodon/blob/main/app/models/concerns/user/omniauthable.rb#L32. If no such user exists, it will generate one. See https://github.com/mastodon/mastodon/blob/main/app/models/concerns/user/omniauthable.rb#L33 and https://github.com/mastodon/mastodon/blob/main/app/models/concerns/user/omniauthable.rb#L66.

The problem is that `OIDC_UID_FIELD` is used for two very different purposes: it defines the attribute that links SSO Identities to mastodon user accounts, and also the default username for new accounts.

**Comment by ThisIsMissEm** [comment]
I'll take another look at this code, maybe I missed something in how it works. If adding a OIDC_USERNAME_FIELD makes sense, I'll open a pull request

**Comment by ThisIsMissEm** [comment]
I've been looking into this, and one of the issues we're going to have is that the code that figures out data for the user/account creation is separate from the way we get `uid` (bare with me, I'm dealing with a migraine so this mightn't make much sense)

Essentially, `auth.uid` is a standard part of all omniauth strategies, where as different strategies have a different mechanism to indicate the "username" field. omniauth-cas has `CAS_NICKNAME_KEY` for `nickname_key` option, which is used when [building the user_info hash](https://github.com/dlindahl/omniauth-cas/blob/master/lib/omniauth/strategies/cas.rb#L58), openid-connect [only has nickname as preferred_username](https://github.com/omniauth/omniauth_openid_connect/blob/23d4b13e67b1e99006d4ec7d0813e8fe64303a9f/lib/omniauth/strategies/openid_connect.rb#L84) (likely because this is what OIDC dictates), omniauth-saml [builds the user_info hash](https://github.com/omniauth/omniauth-saml/blob/master/lib/omniauth/strategies/saml.rb#L102-L109) from the configured `saml_options[:attribute_statements]` key/values, and we do not currently have one for username/nickname/preferred_username, but it'd look like:

```
saml_options[:attribute_statements][:nickname] = [ENV['SAML_ATTRIBUTES_STATEMENTS_NICKNAME']] if ENV['SAML_ATTRIBUTES_STATEMENTS_NICKNAME']
```

Given all that, I think you'd probably want to rely exclusively on `user_info.nickname` for the username property when creating the user/account, moving away from using `auth.uid`; I don't know if making this configurable would make sense, given how the data access is on the `user_info` hash.

So I don't think adding a `OIDC_USERNAME_FIELD` alone would be the right approach, but instead we'd probably want to rely on the `user_info.nickname` field, unless we want to dig into the mess that likely is `raw_info` which will likely vary between providers.
